File: impl/ruby/test/test_bounds.rb
# Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
#
# Decoder resource bounds (design.md §3.4, R7) for the Ruby SDK: the object octet-size ceiling,
# the CBOR nesting-depth ceiling, the causes[]/ext/cext cardinality ceilings, and the stream
# chunk-count ceiling. Mirrors impl/go/cbor/bounds_test.go, impl/go/envelope/bounds_test.go, and
# impl/go/streaming/bounds_test.go (cross-read against impl/rust/src/{cbor,envelope,streaming}.rs).
#
# Every bound is proven by a BOUNDARY PAIR: an otherwise-valid input exactly AT the limit is
# accepted, and one just past the limit is rejected with its exact error kind. "Otherwise valid"
# is load-bearing for mutation survival -- the only defect is the bound, so deleting the bound
# check would make the over-limit case verify instead of reject.
#
# Run:  ruby -Ilib -Itest test/test_bounds.rb      (from impl/ruby/)
require 'minitest/autorun'
require 'naalp'

include Naalp::CBOR

BOUNDS_ALG = Naalp::COSE::ALG_MLDSA65
BOUNDS_SIGNER = "SIGNER_BOUNDS".b

MAXD = Naalp::Envelope::MAX_NESTING_DEPTH
MAXC = Naalp::Envelope::MAX_CAUSES
MAXE = Naalp::Envelope::MAX_EXT
MAXX = Naalp::Envelope::MAX_CEXT
MAXO = Naalp::Envelope::MAX_OBJECT_SIZE
MAXS = Naalp::Envelope::MAX_STREAM_CHUNKS

# deterministic ML-DSA needs OpenSSL >= 3.5; skip loudly (never a false green) where unavailable.
def bounds_mldsa_key
  seed = ("\x07" * 32).b
  pk = Naalp::COSE.mldsa_keygen("ML-DSA-65", seed)
  [seed, pk]
rescue Exception
  nil
end

def bounds_accept_kind
  ->(_ch, _k) { true }
end

# An otherwise-valid object (kind/channel/tier/signer/created/effect/profile all in range); only
# causes/ext/cext/body vary per bound under test.
def bounds_object(causes: [], ext: nil, cext: nil, body: T.new("bounds"))
  Naalp.object(
    kind: 1, channel: 4, tier: 0, signer: BOUNDS_SIGNER, created: 1785000000000,
    effect: 2, profile: Naalp::COSE::PROFILE_PUBLIC, causes: causes, ext: ext, cext: cext, body: body
  )
end

# n content-id-shaped bstrs (multihash sha2-384 = 0x20 0x30 + 48 bytes), like a real causes entry.
def bounds_causes(n)
  Array.new(n) do
    b = ("\x00" * 50).b
    b.setbyte(0, 0x20)
    b.setbyte(1, 0x30)
    b
  end
end

# n distinct non-critical extension entries (unknown keys, which the may-ignore rule accepts), so
# the object is otherwise valid at any cardinality.
def bounds_ext_map(n)
  M.new(Array.new(n) { |i| [U.new(100 + i), U.new(0)] })
end

# k single-element arrays wrapping a zero scalar, as a FIELD_BODY value. The object body map
# decodes at depth 1, so the FIELD_BODY map-entry value sits at depth 2, and the innermost scalar
# sits at depth 2+k (mirrors impl/go/envelope/bounds_test.go nestArrays).
def bounds_nest_arrays(k)
  v = U.new(0)
  k.times { v = A.new([v]) }
  v
end

# ---- C1 CBOR nesting depth (design.md §3.4, R7) --------------------------------------------

class CborBoundsTest < Minitest::Test
  # k single-element arrays (0x81 repeated) wrapping a zero scalar: decoding it, the outermost
  # array is at depth 1 and the innermost scalar is at depth k+1.
  def nested_arrays_cbor(k)
    ([0x81].pack("C") * k) + [0x00].pack("C")
  end

  # Pins the nesting-depth counter (design.md §3.4, R7): the outermost item is depth 1, and
  # decode_bounded rejects the first item at depth max_depth+1 with DepthExceeded, before it is
  # materialized. The unbounded decode path still accepts the same structure, so the bound, not
  # another check, is what does the work.
  def test_decode_bounded_depth
    d = 3

    # deepest scalar at depth d (k = d-1): accepted at max_depth=d.
    at_limit = nested_arrays_cbor(d - 1)
    Naalp::CBOR.decode_bounded(at_limit, d)

    # deepest scalar at depth d+1 (k = d): rejected DepthExceeded at max_depth=d.
    over = nested_arrays_cbor(d)
    err = assert_raises(Naalp::CBOR::DepthExceeded) { Naalp::CBOR.decode_bounded(over, d) }
    assert_equal "DepthExceeded", err.kind

    # the unbounded path accepts the same over-depth structure: the bound above, not another
    # check, is what rejected it.
    Naalp::CBOR.decode(over)
  end
end

# ---- C3 envelope bounds (design.md §3.4, R7) ------------------------------------------------

# An object exactly AT each cardinality/depth bound verifies, so the boundary is inclusive and
# the reject tests below prove the boundary itself.
class EnvelopeBoundsAcceptTest < Minitest::Test
  def setup
    @seed, @pk = bounds_mldsa_key
    skip "deterministic ML-DSA unavailable (needs OpenSSL >= 3.5)" if @pk.nil?
  end

  def accept(name, obj)
    signed = Naalp.sign(obj, BOUNDS_ALG, @seed)
    Naalp.verify(Naalp::COSE::PROFILE_PUBLIC, BOUNDS_ALG, @pk, bounds_accept_kind, signed)
  rescue StandardError => e
    flunk "#{name} verify at-limit: #{e.class}: #{e.message}"
  end

  def test_causes_at_limit
    accept("causes==MAX_CAUSES", bounds_object(causes: bounds_causes(MAXC)))
  end

  def test_ext_at_limit
    accept("ext==MAX_EXT", bounds_object(ext: bounds_ext_map(MAXE)))
  end

  def test_depth_at_limit
    # body nested so the deepest scalar sits at exactly MAX_NESTING_DEPTH (2 + (MAXD-2)).
    accept("depth==MAX_NESTING_DEPTH", bounds_object(body: bounds_nest_arrays(MAXD - 2)))
  end
end

# An otherwise-valid object one past each bound is rejected with its named error (fail-closed).
class EnvelopeBoundsRejectTest < Minitest::Test
  def setup
    @seed, @pk = bounds_mldsa_key
    skip "deterministic ML-DSA unavailable (needs OpenSSL >= 3.5)" if @pk.nil?
  end

  def expect(name, obj, error_class, kind)
    signed = Naalp.sign(obj, BOUNDS_ALG, @seed)
    err = assert_raises(error_class) do
      Naalp.verify(Naalp::COSE::PROFILE_PUBLIC, BOUNDS_ALG, @pk, bounds_accept_kind, signed)
    end
    assert_equal kind, err.kind, name
  end

  def test_too_many_causes
    expect("TooManyCauses", bounds_object(causes: bounds_causes(MAXC + 1)),
           Naalp::Envelope::EnvelopeError, "TooManyCauses")
  end

  def test_too_many_extensions_ext
    expect("TooManyExtensions(ext)", bounds_object(ext: bounds_ext_map(MAXE + 1)),
           Naalp::Envelope::EnvelopeError, "TooManyExtensions")
  end

  # cext over the limit also yields TooManyExtensions: the cardinality check in object_from_map
  # fires before the critical-extension recognition check.
  def test_too_many_extensions_cext
    expect("TooManyExtensions(cext)", bounds_object(cext: bounds_ext_map(MAXX + 1)),
           Naalp::Envelope::EnvelopeError, "TooManyExtensions")
  end

  def test_depth_exceeded
    # body nested so the deepest scalar sits at MAX_NESTING_DEPTH+1.
    expect("DepthExceeded", bounds_object(body: bounds_nest_arrays(MAXD - 1)),
           Naalp::CBOR::DepthExceeded, "DepthExceeded")
  end
end

# Pins the object octet-size bound: a large-but-under-limit signed object verifies, and an
# otherwise-valid object over the limit is rejected TooLarge on the raw bytes before any parse
# (RFC 8949 §10 decoder-memory guard).
class EnvelopeBoundsTooLargeTest < Minitest::Test
  def setup
    @seed, @pk = bounds_mldsa_key
    skip "deterministic ML-DSA unavailable (needs OpenSSL >= 3.5)" if @pk.nil?
  end

  def test_too_large
    under = bounds_object(body: B.new(("\x00" * (MAXO - 16384)).b))
    uobj = Naalp.sign(under, BOUNDS_ALG, @seed)
    assert uobj.bytesize < MAXO, "under-limit object is #{uobj.bytesize} bytes, expected < #{MAXO}"
    Naalp.verify(Naalp::COSE::PROFILE_PUBLIC, BOUNDS_ALG, @pk, bounds_accept_kind, uobj)

    over = bounds_object(body: B.new(("\x00" * MAXO).b))
    bobj = Naalp.sign(over, BOUNDS_ALG, @seed)
    assert bobj.bytesize > MAXO, "over-limit object is only #{bobj.bytesize} bytes, expected > #{MAXO}"
    err = assert_raises(Naalp::Envelope::EnvelopeError) do
      Naalp.verify(Naalp::COSE::PROFILE_PUBLIC, BOUNDS_ALG, @pk, bounds_accept_kind, bobj)
    end
    assert_equal "TooLarge", err.kind
  end
end

# ---- C9 streaming chunk-count bound (design.md §3.4, R7) ------------------------------------

# A commit/checkpoint over exactly MAX_STREAM_CHUNKS chunks verifies, and one over
# MAX_STREAM_CHUNKS+1 is rejected TooManyChunks. Both carry a MATCHING rolling digest, so the
# count is the only reason to reject -- deleting the count check would make the +1 case verify
# (the mutation is caught). The chunks are empty so building/hashing a million-plus of them stays
# cheap.
class StreamingBoundsTest < Minitest::Test
  def test_bound_too_many_chunks
    max = MAXS
    over = Array.new(max + 1) { Naalp::Streaming::Chunk.new(0, "".b) }
    at_limit = over[0, max]

    ok_commit = Naalp::Streaming::StreamCommit.new("".b, Naalp::Streaming.commit_digest(at_limit))
    Naalp::Streaming.verify_commit(ok_commit, at_limit)

    over_commit = Naalp::Streaming::StreamCommit.new("".b, Naalp::Streaming.commit_digest(over))
    err = assert_raises(Naalp::Streaming::StreamError) { Naalp::Streaming.verify_commit(over_commit, over) }
    assert_equal "TooManyChunks", err.kind

    # verify_checkpoint enforces the same bound, and the count check fires before the
    # contiguity/digest checks, so the diagnosis is TooManyChunks (not a digest error).
    cp = Naalp::Streaming::StreamCheckpoint.new("".b, 0, "".b)
    cerr = assert_raises(Naalp::Streaming::StreamError) { Naalp::Streaming.verify_checkpoint(cp, over) }
    assert_equal "TooManyChunks", cerr.kind
  end
end

File: impl/python/tests/test_bounds.py
# Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
"""
Decoder resource bounds (design.md §3.4, R7) known-answer + fail-closed tests for the Python
SDK, mirroring impl/go/cbor/bounds_test.go, impl/go/envelope/bounds_test.go, and
impl/go/streaming/bounds_test.go.

Each bound is proven by a BOUNDARY PAIR: an otherwise-valid input exactly AT the limit
verifies, and an otherwise-valid input one past the limit is rejected with the named error.
"Otherwise valid" is load-bearing for mutation survival -- because the only defect is the
bound, deleting the bound check makes the over-limit input verify.

Run:  python -m unittest -v tests.test_bounds   (from impl/python/, PYTHONDONTWRITEBYTECODE=1)
"""
import unittest

from naalp import cbor, cose, envelope, streaming
from naalp.cbor import U, B, T, A, M

_ALG = cose.ALG_MLDSA65
_SEED = bytes(range(32))  # a LOCAL test signing seed -- the verdict is a real sign+verify.
_SIGNER = b"BOUNDS_SIGNER"


def _kind_ok(_ch, _k):
    return True


def _base_object(body=None, causes=None, ext=None, cext=None):
    """A minimal otherwise-valid object (channel 0, kind 1 -- not the Identity Rotation
    selector) so the only defect under test is the bound itself."""
    return envelope.Object(
        kind=1, channel=0, signer=_SIGNER, created=1785000000000, effect=0,
        body=body if body is not None else T("bounds"),
        tier=0, profile=cose.PROFILE_PUBLIC, causes=causes or [], ext=ext, cext=cext,
    )


def _make_causes(n):
    """n content-id-shaped bstrs (multihash sha2-384 = 0x20 0x30 + 48 zero bytes)."""
    out = []
    for _ in range(n):
        b = bytearray(50)
        b[0], b[1] = 0x20, 0x30
        out.append(bytes(b))
    return out


def _make_ext_map(n):
    """n distinct non-critical/unrecognized extension entries (unknown keys, which the
    may-ignore rule accepts), so the object is otherwise valid at any cardinality."""
    return M([(U(100 + i), U(0)) for i in range(n)])


def _nest_arrays(k):
    """k single-element arrays wrapping a zero scalar. As a body value it sits at depth 2 (the
    object body map is depth 1), so the scalar is at depth 2+k."""
    v = U(0)
    for _ in range(k):
        v = A([v])
    return v


# ---- CBOR-level nesting depth (mirrors impl/go/cbor/bounds_test.go) --------------------------

def _nested_arrays_cbor(k):
    """Canonical CBOR for k single-element arrays wrapping a zero scalar: 0x81 (array of one)
    repeated k times, then 0x00. Decoding it, the outermost array is at depth 1 and the
    innermost scalar is at depth k+1."""
    return bytes([0x81]) * k + bytes([0x00])


class DecodeBoundedDepthTest(unittest.TestCase):
    """Pins the nesting-depth counter (design.md §3.4, R7): the outermost item is depth 1, and
    decode_bounded rejects the first item at depth max_depth+1 with DepthExceeded, before it is
    materialized. The unbounded decode() path still accepts the same structure, so the bound is
    what does the work."""

    def test_decode_bounded_depth(self):
        d = 3

        # deepest scalar at depth d (k = d-1): accepted at max_depth=d.
        at_limit = _nested_arrays_cbor(d - 1)
        cbor.decode_bounded(at_limit, d)  # must not raise

        # deepest scalar at depth d+1 (k = d): rejected DepthExceeded at max_depth=d.
        over = _nested_arrays_cbor(d)
        with self.assertRaises(cbor.DepthExceeded) as cm:
            cbor.decode_bounded(over, d)
        self.assertEqual(cm.exception.kind, "DepthExceeded")

        # the unbounded path accepts the same over-depth structure: the bound, not another
        # check, is what rejected it above.
        cbor.decode(over)  # must not raise


# ---- Envelope-level bounds (mirrors impl/go/envelope/bounds_test.go) --------------------------

class EnvelopeBoundsAcceptAtLimitTest(unittest.TestCase):
    """An object exactly AT each cardinality/depth bound verifies, so the boundary is inclusive
    and the reject tests below prove the boundary itself."""

    def test_causes_at_limit(self):
        pk = cose.mldsa_keygen("ML-DSA-65", _SEED)
        o = _base_object(causes=_make_causes(envelope.MAX_CAUSES))
        signed = envelope.sign(o, _ALG, _SEED)
        envelope.verify(cose.PROFILE_PUBLIC, _ALG, pk, _kind_ok, signed)  # must not raise

    def test_ext_at_limit(self):
        pk = cose.mldsa_keygen("ML-DSA-65", _SEED)
        o = _base_object(ext=_make_ext_map(envelope.MAX_EXT))
        signed = envelope.sign(o, _ALG, _SEED)
        envelope.verify(cose.PROFILE_PUBLIC, _ALG, pk, _kind_ok, signed)  # must not raise

    def test_depth_at_limit(self):
        pk = cose.mldsa_keygen("ML-DSA-65", _SEED)
        # body nested so the deepest scalar sits at exactly MAX_NESTING_DEPTH
        # (2 + (MAX_NESTING_DEPTH-2)).
        o = _base_object(body=_nest_arrays(envelope.MAX_NESTING_DEPTH - 2))
        signed = envelope.sign(o, _ALG, _SEED)
        envelope.verify(cose.PROFILE_PUBLIC, _ALG, pk, _kind_ok, signed)  # must not raise


class EnvelopeBoundsRejectOverLimitTest(unittest.TestCase):
    """An otherwise-valid object one past each bound is rejected with its named error
    (fail-closed)."""

    def _expect(self, obj, kind):
        pk = cose.mldsa_keygen("ML-DSA-65", _SEED)
        signed = envelope.sign(obj, _ALG, _SEED)
        with self.assertRaises(Exception) as cm:
            envelope.verify(cose.PROFILE_PUBLIC, _ALG, pk, _kind_ok, signed)
        self.assertEqual(getattr(cm.exception, "kind", None), kind)

    def test_causes_over_limit(self):
        self._expect(_base_object(causes=_make_causes(envelope.MAX_CAUSES + 1)), "TooManyCauses")

    def test_ext_over_limit(self):
        self._expect(_base_object(ext=_make_ext_map(envelope.MAX_EXT + 1)), "TooManyExtensions")

    def test_cext_over_limit(self):
        # cext over the limit also yields TooManyExtensions: the cardinality check in
        # _object_from_map fires before the critical-extension recognition check.
        self._expect(_base_object(cext=_make_ext_map(envelope.MAX_CEXT + 1)), "TooManyExtensions")

    def test_depth_over_limit(self):
        # body nested so the deepest scalar sits at MAX_NESTING_DEPTH+1.
        self._expect(_base_object(body=_nest_arrays(envelope.MAX_NESTING_DEPTH - 1)), "DepthExceeded")


class EnvelopeBoundTooLargeTest(unittest.TestCase):
    """Pins the object octet-size bound: a large-but-under-limit signed object verifies, and an
    otherwise-valid object over the limit is rejected TooLarge on the raw bytes before any parse
    (RFC 8949 §10 decoder-memory guard)."""

    def test_bound_too_large(self):
        pk = cose.mldsa_keygen("ML-DSA-65", _SEED)

        under = _base_object(body=B(bytes(envelope.MAX_OBJECT_SIZE - 16384)))
        uobj = envelope.sign(under, _ALG, _SEED)
        self.assertLessEqual(len(uobj), envelope.MAX_OBJECT_SIZE,
                              "under-limit object exceeded MAX_OBJECT_SIZE")
        envelope.verify(cose.PROFILE_PUBLIC, _ALG, pk, _kind_ok, uobj)  # must not raise

        over = _base_object(body=B(bytes(envelope.MAX_OBJECT_SIZE)))
        bobj = envelope.sign(over, _ALG, _SEED)
        self.assertGreater(len(bobj), envelope.MAX_OBJECT_SIZE,
                            "over-limit object did not exceed MAX_OBJECT_SIZE")
        with self.assertRaises(envelope.EnvelopeError) as cm:
            envelope.verify(cose.PROFILE_PUBLIC, _ALG, pk, _kind_ok, bobj)
        self.assertEqual(cm.exception.kind, "TooLarge")


# ---- Streaming chunk-count bound (mirrors impl/go/streaming/bounds_test.go) -------------------

class StreamBoundTooManyChunksTest(unittest.TestCase):
    """Pins the stream chunk-count bound (design.md §3.4, R7): a commit over exactly
    MAX_STREAM_CHUNKS chunks verifies, and one over MAX_STREAM_CHUNKS+1 is rejected
    TooManyChunks. Both carry a MATCHING rolling digest, so the count is the only reason to
    reject -- deleting the count check makes the +1 case verify (the mutation is caught). The
    chunks are all-empty so the accept-at-limit case is built cheaply."""

    def test_bound_too_many_chunks(self):
        over = [streaming.Chunk(offset=0, data=b"") for _ in range(streaming.MAX_STREAM_CHUNKS + 1)]
        at_limit = over[:streaming.MAX_STREAM_CHUNKS]

        ok_commit = streaming.StreamCommit(stream_id=b"s", digest=streaming.commit_digest(at_limit))
        streaming.verify_commit(ok_commit, at_limit)  # must not raise

        over_commit = streaming.StreamCommit(stream_id=b"s", digest=streaming.commit_digest(over))
        with self.assertRaises(streaming.StreamError) as cm:
            streaming.verify_commit(over_commit, over)
        self.assertEqual(cm.exception.kind, "TooManyChunks")

        # verify_checkpoint enforces the same bound, and the count check fires before the
        # contiguity/digest checks, so the diagnosis is TooManyChunks (not a digest error).
        over_cp = streaming.StreamCheckpoint(stream_id=b"s", through_offset=0, digest_so_far=b"")
        with self.assertRaises(streaming.StreamError) as cm2:
            streaming.verify_checkpoint(over_cp, over)
        self.assertEqual(cm2.exception.kind, "TooManyChunks")


if __name__ == "__main__":
    unittest.main()
